src/compiler/scala/tools/nsc/transform/TailCalls.scala

```diff
@@ -116,7 +116,9 @@ abstract class TailCalls extends Transform {
       def tailLabels: Set[Symbol]
 
       def enclosingType = method.enclClass.typeOfThis
-      def isEligible    = method.isEffectivelyFinalOrNotOverridden
+      def isEligible    = method.isEffectivelyFinalOrNotOverridden && !isExcludedBySingletonArg && !isExcludedBySingletonResult
+      def isExcludedBySingletonArg = method.paramss.exists(_.exists(_.tpe_*.isInstanceOf[SingletonType]))
+      def isExcludedBySingletonResult = method.tpe_*.finalResultType.isInstanceOf[SingletonType]
```

Comment:
We've got `mexists` for the `exists(_.exists(_ => )` pattern. More utilities like it are with `reflect.internal.util.Collections`.